tools/serverpod_cli/lib/src/analyzer/entities/stateful_analyzer.dart

```diff
@@ -0,0 +1,144 @@
+  /// the state. Subsequent validations should use [validateAll] or [validateProtocol].
+  List<SerializableEntityDefinition> initialValidation(
+      List<ProtocolSource> sources) {
+    for (var yamlSource in sources) {
+      var analyzer = SerializableEntityAnalyzer(
+        yaml: yamlSource.yaml,
+        sourceFileName: yamlSource.uri.path,
+        subDirectoryParts: yamlSource.protocolRootPathParts,
+        collector: CodeGenerationCollector(),
+      );
+
+      _analyzers[yamlSource.uri.path] = analyzer;
+    }
+
+    _entities = _validateAllAnalyzers();
+    _entities = _validateAllAnalyzers(_entities);
+    return _entities;
+  }
+
+  /// Runs the validation on all protocols, assumes [initialValidation] has been
+  /// run before, if not this returns an empty list.
+  /// Errors are reported through the [onErrorsChangedNotifier].
+  List<SerializableEntityDefinition> validateAll() {
+    _entities = _validateAllAnalyzers(_entities);
+    return _entities;
+  }
+
+  /// Runs the validation on a single protocol, assumes [initialValidation] has been
+  /// run before. The protocol must exist in the state, if not this returns
+  /// the last validated state. To validate a new protocol, use [addYamlProtocol]
+  /// and then run [validateAll] or [validateProtocol].
+  /// Errors are reported through the [onErrorsChangedNotifier].
+  List<SerializableEntityDefinition> validateProtocol(String yaml, Uri uri) {
+    var analyzer = _analyzers[uri.path];
+    if (analyzer == null) return _entities;
+
+    analyzer.collector.clearErrors();
+    var document = analyzer.analyze(yaml: yaml, protocolEntities: _entities);
+    if (document != null) {
+      _upsertEntity(document, uri);
+      // TODO if inserted, validate all again or we may miss errors.
+      // for now we let the caller take care of this.
+    }
+
+    _onErrorsChangedNotifier?.call(
+      uri,
+      (analyzer.collector as CodeGenerationCollector),
+    );
+
+    return _entities;
+  }
+
+  /// Adds a new protocol to the state but leaves the responsibility of validating
+  /// it to the caller. Please note that the validation needs to be done twice
+  /// for the first validation pass to detect all errors.
+  void addYamlProtocol(ProtocolSource yamlSource) {
+    var analyzer = SerializableEntityAnalyzer(
+      yaml: yamlSource.yaml,
+      sourceFileName: yamlSource.uri.path,
+      subDirectoryParts: yamlSource.protocolRootPathParts,
+      collector: CodeGenerationCollector(),
+    );
+
+    _analyzers[yamlSource.uri.path] = analyzer;
+  }
+
+  /// Removes a protocol from the state but leaves the responsibility of validating
+  /// the new state to the caller.
+  void removeYamlProtocol(Uri protocolUri) {
+    _analyzers.remove(protocolUri.path);
+    _entities
+        .removeWhere((entity) => entity.sourceFileName == protocolUri.path);
+  }
+
+  /// Register a callback that is called when the errors in a file changes.
+  void regsiterOnErrorsChangedNotifier(
```

Comment:
NIT: Spelling
```suggestion
  void registerOnErrorsChangedNotifier(
```

But hopefully it can be removed.